clang/lib/Format/WhitespaceManager.cpp

```diff
@@ -14,9 +14,81 @@
 #include "WhitespaceManager.h"
 #include "llvm/ADT/STLExtras.h"
 
+#include <fstream>
+
 namespace clang {
 namespace format {
 
+class Logger {
+public:
+    static Logger& Get() {
+        static Logger instance;
+        return instance;
+    }
+
+    void Init(const std::string &NewFileName) {
+        if (FileName != NewFileName) {
+            LogStream.open(NewFileName.c_str());
+            FileName = NewFileName;
+        }
+    }
+
+    int Indent(int delta = 0) {
+        return CurrentIndent += delta;
+    }
+
+private:
+    Logger() : CurrentIndent(0) {}
+
+    friend class LogLine;
+    std::string   FileName;
+    std::ofstream LogStream;
+    int           CurrentIndent;
+};
+
+class LogLine {
+public:
+    LogLine(const char *File, int Line) : Log(Logger::Get()) {
+        if (!Log.FileName.empty()) {
+            Log.LogStream << "[CLANG] " << basename(File) << ":" << Line << "   ";
+            for (int i = 0, e = Log.Indent(); i < e; ++i) {
+                Log.LogStream << " ";
+            }
+        }
+    }
+
+    ~LogLine() {
+        if (!Log.FileName.empty()) {
+            Log.LogStream << std::endl;
+        }
+    }
+
+    template<typename T>
+    const LogLine& operator<<(const T &t) const {
+        if (!Log.FileName.empty()) {
+            Log.LogStream << t;
+        }
+        return *this;
+    }
+
+private:
+    Logger &Log;
+};
+
+class LogScope {
+public:
+    LogScope() {
+        Logger::Get().Indent(2);
+    }
```

Comment:
Another suggestion: Put the `2` constant in `Logger`, and only expose `Indent()`, `Unindent()` methods (w/o any args), which `LogScope` will use.

Then, `LogScope` doesn't need to know "how much is a single indent", and only makes a single "step" each time, and the indent size is an internal implementation detail in `Logger`.